### DISCUSSION TITLE
Ubernerd: Portable install of nerdctl (without modifying host rootfs)

### DISCUSSION THREAD

**Discussion by Jip-Hop** [containerd/nerdctl#2460] [Show and tell]
Hi all, I'm developing [an experimental install script](https://github.com/Jip-Hop/ubernerd) which allows starting `containerd` and running `nerdctl` without modifying the host rootfs. I'm trying to reduce the possibility for conflicts as much as possible. Looking forward to your feedback. 🙂

**Comment by fahedouch** [comment]
Hi @Jip-Hop, 

thanks for you effort. It would be nice if the portable install of nerdctl/containerd could provide more isolation such as a  container isolation ( e.g by mounting config and data folder in a  container including containerd sock ) , this may help many people asking for containerd-in-containerd https://github.com/containerd/nerdctl/discussions/2008 https://github.com/containerd/nerdctl/discussions/1383

I gave you the first star ;)

**Reply by Jip-Hop** [reply]
Yes, it would be cool if the portable install would simplify/allow containerd in containerd. Would require updates to the script though, as right now it starts containerd using systemd (which isn't available by default in most container images).

In the mean time I'm running into [this 'hint' from nerdctl](https://github.com/Jip-Hop/ubernerd/issues/2): "try running with `dhcp daemon --socketpath=/run/cni/dhcp.sock &` in CNI_PATH to launch the dhcp daemon". But `CNI_PATH` seems to be an environment variable... So I don't see how putting these commands inside an environment variable would enable the dhcp daemon.

Do you have any idea? Would be great to also start the dhcp daemon automatically.

**Reply by fahedouch** [reply]
> Alternatively, you can use systemd socket activation protocol. Be sure that the .socket file uses /run/cni/dhcp.sock as the socket path.

I would recommend `systemd socket activation protocol`  to create `dhcp` socket for `Ubernerd` but unfortunately, they enforce `/run/cni/dhcp.sock ` as socket path which is not a good choice for `Ubernerd` case ( may be it works with other paths it deserves a test)

> "try running with dhcp daemon --socketpath=/run/cni/dhcp.sock & in CNI_PATH to launch the dhcp daemon".

I agree, the warn message is not clear, IMO it is  just asking to launch `dhcp` daemon binary under `CNI_PATH`

What I recommend in your case is : 
```
dhcp daemon --socketpath=/path/to/data/dir
```

**Reply by Jip-Hop** [reply]
On a Debian 12 host I can run `nerdctl` inside a container using the containerd socket on the host:

```sh
nerdctl run --privileged --rm --net=host -v /root/ubern:/root/ubern -v /run/ubernerd/containerd:/run/ubernerd/containerd -v /run/containerd:/run/containerd -v /etc/ssl/certs:/etc/ssl/certs:ro -w /root/ubern debian ./nerdctl run hello-world
```

The above runs successfully after running `ubernerd.sh` inside the `/root/ubern` directory I created for it.

Is this what you meant?

> It would be nice if the portable install of nerdctl/containerd could provide more isolation such as a container isolation ( e.g by mounting config and data folder in a container including containerd sock )

Slight mystery still as to why `/run/containerd` is still being used (in addition to `/run/ubernerd/containerd`). I thought I changed all the relevant config to `/run/ubernerd/containerd`...

**Reply by Jip-Hop** [reply]
Regarding installing `ubernerd` and running `nerdctl` inside a container (without connecting to the `containerd` socket on the host). It's only working for me when I specify `--snapshotter=native`. Otherwise I get:

> FATA[0000] failed to create shim task: failed to mount rootfs component: invalid argument: unknown 

Tried to run inside a container created with `docker run`, as well as `nerdctl run` inside `nerdctl run`. An example can be found here: https://github.com/Jip-Hop/ubernerd/blob/main/test.sh#L22-L30

**Reply by fahedouch** [reply]
> Is this what you meant?

Yes. but I expected a lot less of container mounts

>. Slight mystery still as to why /run/containerd is still being used (in addition to /run/ubernerd/containerd). I thought I changed all the relevant config to /run/ubernerd/containerd...

strange, are you facing some errors without  `/run/containerd` ?

> Regarding installing ubernerd and running nerdctl inside a container (without connecting to the containerd socket on the host). It's only working for me when I specify --snapshotter=native. Otherwise I get:

IMO `ubernerd` should focus on mount host containerd sock provided by  `ubernerd.sh` setup. We already provider a containerd setup inside a [docker image ](https://github.com/containerd/nerdctl/blob/main/Dockerfile)

**Reply by Jip-Hop** [reply]
Without mounting `/run/containerd` I get:
> time="2023-09-01T07:36:20Z" level=fatal msg="failed to create shim task: failed to start io pipe copy: unable to copy pipes: containerd-shim: opening file \"/run/containerd/fifo/3297564989/a1084e5bef3f95e71a08a0d3b9e397534dd50b6d2773f7c2865d2e1f350c5d76-stdout\" failed: open /run/containerd/fifo/3297564989/a1084e5bef3f95e71a08a0d3b9e397534dd50b6d2773f7c2865d2e1f350c5d76-stdout: no such file or directory: unknown"

Did I miss some config to change this path from `/run/containerd` to `/run/ubernerd/containerd` or is it not possible to change this path?

**Reply by Jip-Hop** [reply]
> Yes. but I expected a lot less of container mounts

If this fifo dir is configurable I think I can make it work with a single mount.

**Reply by Jip-Hop** [reply]
P.S. [Running `ubernerd.sh` inside a docker container now works without `--snapshotter=native`](https://github.com/Jip-Hop/ubernerd/blob/main/test.sh). The issue I ran into was due to using a bind mounted directory from macOS into the docker container.

**Reply by philrhc** [reply]
i have used this to run containerd-in-containerd connecting to the host containerd.sock, the script uses the version from https://api.github.com/repos/containerd/nerdctl/releases/latest which is 1.7.6

@Jip-Hop this is really good work, thank you

@fahedouch 
  > this may help many people asking for containerd-in-containerd #2008 #1383  

i also think this will help people

**Comment by ningziwen** [comment]
Thanks for the effort! Gave the 3rd star :)

Going through the scripts and documentation, the use case seems pretty similar to Finch Linux (tracking [issue](https://github.com/runfinch/finch/issues/377)).

Looking forward to the potential collaboration if the use cases can align.

**Reply by Jip-Hop** [reply]
Looks interesting! But idk if Finch Linux would be as portable as I'd like ubernerd to be. Not sure if that's the intention for Fine Linux, but I definitely don't want to run nerdctl inside a VM.

**Reply by ningziwen** [reply]
Yeah. Finch macOS and Windows have to use VM, but whether to use VM in Finch Linux is still open. We have not discussed much about it yet. 

Considering the nested VM issue of running Finch Linux in cloud instance such as EC2, there are definitely reasons supporting no-VM portable Finch Linux.

Feel free to share your thoughts in Finch Linux issue. The input from you or other community members will contribute to the decision of whether to go.
